armadillo/src/main/java/org/molgenis/armadillo/controller/StorageController.java

```diff
@@ -0,0 +1,292 @@
+
+  public StorageController(ArmadilloStorageService storage, AuditEventPublisher auditor) {
+    this.storage = storage;
+    this.auditor = auditor;
+  }
+
+  @Operation(summary = "List all projects")
+  @ApiResponses(
+      value = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "Projects listed",
+            content =
+                @Content(array = @ArraySchema(schema = @Schema(implementation = String.class)))),
+        @ApiResponse(
+            responseCode = "401",
+            description = "Unauthorized",
+            content = @Content(schema = @Schema(hidden = true)))
+      })
+  @GetMapping(
+      value = "/projects",
+      produces = {APPLICATION_JSON_VALUE})
+  @ResponseStatus(OK)
+  public List<String> listProjects(Principal principal) {
+    return auditor.audit(storage::listProjects, principal, LIST_PROJECTS, Map.of());
+  }
+
+  @Operation(summary = "Project exists?")
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "Project exists"),
+        @ApiResponse(responseCode = "404", description = "Project does not exist"),
+        @ApiResponse(responseCode = "401", description = "Unauthorized")
+      })
+  @RequestMapping(value = "/projects/{project}", method = HEAD)
+  public ResponseEntity<Void> projectExists(Principal principal, @PathVariable String project) {
+    boolean projectExists =
+        auditor.audit(
+            () -> storage.hasProject(project), principal, GET_PROJECT, Map.of(PROJECT, project));
+    return projectExists ? noContent().build() : notFound().build();
+  }
+
+  @Operation(summary = "List objects in a project")
+  @ApiResponses(
+      value = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "Objects listed",
+            content =
+                @Content(array = @ArraySchema(schema = @Schema(implementation = String.class)))),
+        @ApiResponse(
+            responseCode = "404",
+            description = "Project does not exist",
+            content = @Content(schema = @Schema(hidden = true))),
+        @ApiResponse(
+            responseCode = "401",
+            description = "Unauthorized",
+            content = @Content(schema = @Schema(hidden = true)))
+      })
+  @GetMapping(
+      value = "/projects/{project}/objects",
+      produces = {APPLICATION_JSON_VALUE})
+  @ResponseStatus(OK)
+  public List<String> listObjects(Principal principal, @PathVariable String project) {
+    return auditor.audit(
+        () -> storage.listObjects(project), principal, LIST_OBJECTS, Map.of(PROJECT, project));
+  }
+
+  @Operation(summary = "Upload an object to a project")
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "Object uploaded successfully"),
+        @ApiResponse(responseCode = "404", description = "Unknown project"),
+        @ApiResponse(responseCode = "409", description = "Object already exists"),
```

Comment:
we could consider to convert from 'post' to 'put'. Then you would overwrite existing. Might be more convenient? In the admin API we implemented this upsert pattern (which made much sense in light that we there really update a json blob we had to have update functionality)

Reply:
Hmm, I haven't considered it, but it sounds a bit dangerous to me. You could delete/overwrite data you didn't intend to without you knowing about it. I think doing one extra request is worth the added security.

Reply:
ok

Reply:
ok